sdk/src/main/java/com/mapbox/maps/renderer/FpsManager.kt

```diff
@@ -0,0 +1,208 @@
+    if (userDefinedRefreshRate != updatedMaxFps) {
+      userDefinedRefreshRate = updatedMaxFps
+      logI(TAG, "User set max FPS to $userDefinedRefreshRate")
+      if (screenRefreshRate != SCREEN_METRICS_NOT_DEFINED) {
+        userDefinedRatio = userDefinedRefreshRate.toDouble() / screenRefreshRate
+        logI(TAG, "User defined ratio is $userDefinedRatio")
+      }
+    }
+  }
+
+  @AnyThread
+  fun setFpsChangedListener(fpsChangedListener: OnFpsChangedListener?) {
+    this.fpsChangedListener = fpsChangedListener
+  }
+
+  /**
+   * Return true if rendering should happen this frame and false otherwise.
+   *
+   * @param frameTimeNs time value from Choreographer in [System.nanoTime] timebase.
+   */
+  @WorkerThread
+  fun preRender(frameTimeNs: Long): Boolean {
+    // clear any scheduled task as new render call is about to happen
+    handler?.removeCallbacksAndMessages(null)
+    // no need to perform neither pacing nor FPS calculation when setMaxFps / setOnFpsChangedListener was not called by the user
+    if (userDefinedRatio == USER_DEFINED_RATIO_NOT_SET && fpsChangedListener == null) {
+      return true
+    }
+    preRenderTimeNs = System.nanoTime()
+    var skippedNow = 0
+    // check if we did miss VSYNC deadline meaning too much work was done in previous doFrame
+    if (previousRefreshNs != -1L && frameTimeNs - previousRefreshNs > screenRefreshPeriodNs + ONE_MILLISECOND_NS) {
+      skippedNow =
+        ((frameTimeNs - previousRefreshNs) / (screenRefreshPeriodNs + ONE_MILLISECOND_NS)).toInt()
+      choreographerSkips += skippedNow
+    }
+    previousRefreshNs = frameTimeNs
+    // we always increase choreographer tick by one + add number of skipped frames for consistent results
+    choreographerTicks += skippedNow + 1
+    if (skippedNow > 0 && LOG_STATISTICS) {
+      logW(
+        TAG,
+        "Skipped $skippedNow VSYNC pulses since last actual render," +
+          " total skipped in measurement period $choreographerSkips / $choreographerTicks"
+      )
+    }
+    if (userDefinedRatio != USER_DEFINED_RATIO_NOT_SET) {
+      return performPacing()
+    }
+    return true
+  }
+
+  @WorkerThread
+  fun postRender() {
+    fpsChangedListener?.let { listener ->
+      val frameRenderTimeNs = System.nanoTime() - preRenderTimeNs
+      frameRenderTimeAccumulatedNs += frameRenderTimeNs
+      // normally we update FPS counter once a second
+      if (choreographerTicks >= screenRefreshRate) {
+        calculateFps(listener)
+      } else {
+        // however to produce correct values we also update FPS after IDLE_TIMEOUT_MS
+        // otherwise when updating the map after it was IDLE first update will report
+        // huge delta between new frame and last frame (as we're using dirty rendering)
+        handler?.postDelayed(IDLE_TIMEOUT_MS) {
+          calculateFps(listener)
+          previousRefreshNs = -1
+        }
+      }
+      preRenderTimeNs = -1L
+    }
+  }
+
+  @WorkerThread
+  fun onSurfaceDestroyed() {
```

Comment:
I guess it's not critical but nice to add `handler?.removeCallbacksAndMessages(null)`

Reply:
👍